### PR TITLE
NIFI-9605: Docker Images  for OpenShift

### PR DESCRIPTION
**Pull request by guimou**
Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

The provided Docker images are not easily usable in OpenShift with its built-in security restrictions. To directly use those images you have to create specific Security Context Constraints (SCC) and give further capabilities to Service Accounts just to be able to have initContainers changing file ownership inside the images.
The new container image definitions, based on [ubi8/OpenJDK11](https://catalog.redhat.com/software/containers/ubi8/openjdk-11/5dd6a4b45a13461646f677f4), are fully compatible with OpenShift and have the right files and directories ownership from the start, allowing to use them without relaxing security constraints. That makes them usable by standard users who would not normally have the rights to create those SCCs.
Images definitions are provided for nifi, nifi-registry and nifi-toolkit, in folders named `openshift` alongside the `dockerhub` folders in their respective location.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [ ] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [X] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [X] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] Have you verified that the full build is successful on JDK 8?
- [ ] Have you verified that the full build is successful on JDK 11?
- [X] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [X] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

Signed-off-by: Guillaume Moutier <guillaume.moutier@gmail.com>

### CODE REVIEW

**Comment by joewitt** [comment]
Hello.  Thanks for contributing.  First we will need a JIRA to exist and the commit messsage should link to it.  The JIRA ideally will lay out the overall goal/intent (which appears to be to provide an openshift friendly image for NiFi).  Also, it looks like there are various versions/mechanisms for setting versions.  You'll want everything in maven version land set to 1.16.0-SNAPSHOT which is the current main and then where there need to be hardcoded version references in Dockerfiles and such presumably it would be 1.16.0.  I also see references to much older version of nifi in here.  Best to start fresh with what this would target which at present is 1.16.  So you can/should remove references to older versions of nifi and any associated special handling.

**Comment by guimou** [comment]
Thanks a lot! I've created the Jira issue and changed the title to point to it.
I'll make the changes for 1.16.0, no problem, but are there nightly builds or whatever so that I can test, or do I have to build them myself?

**Comment by guimou** [comment]
Hard-coded versions bumped to 1.16.0 as suggested.

**Comment by kevdoran** [comment]
@guimou There are no nightlies, so you can test with a release version such as 1.15.3. I know that complicates things with the hardcoded version in files such as DockerImage.txt. I'm actually thinking about a new approach so that we don't need those files / hardcoded versions in the code base.

To @joewitt's point though, the idea here is that we are always preparing the source code to be ready for the next release, so if someone downloads the source tarball for 1.16.0 (when it is ready) and builds the Docker images, it will default to fetching those artifacts. During development / testing on main, that does make it trickier. For this PR, if it is working with 1.15.3, I would say that is good for verifying everything is in good order

**Comment by guimou** [comment]
@kevdoran Perfect, thanks a lot. In this case everything has already been tested with 1.15.2 as part of [this project](https://github.com/guimou/odh-manifests/tree/nifi/nifi), both as a simple Kustomize configuration and with the [Open Data Hub](https://opendatahub.io/) operator. Built images are available here: [Nifi](https://quay.io/repository/rh-data-services/nifi-openshift), [Registry](https://quay.io/repository/rh-data-services/nifi-registry-openshift) and [Toolkit](https://quay.io/repository/rh-data-services/nifi-toolkit-openshift).
As 1.15.3 was only a bug fix over 1.15.2 I would say it's pretty much ok, although I will surely find some time to rebuild images with 1.15.3 in the next few days. However it won't change anything for the PR as everything is already set to 1.16.0.

**Comment by dmnewber** [comment]
Hi @guimou, thanks for the PR! It looks like there's a lot of duplication here, particularly with the `sh` directories. It would be better from a maintenance perspective to not have the scripts duplicated and instead enhance the build scripts to be able to select different Dockerfiles. Did you have a goal in mind with duplicating the entire build directory?

**Comment by guimou** [comment]
Hi @dmnewber!
The idea of creating the openshift folders alongside the docker/dockerhub ones (btw naming and structure is not really consistent for those) was to minimize coupling, in the sense that I did not want to "get in the way" (I was not even sure the project would be interested by this). In my idea, people would be able to build the images themselves, or we could have an external automated build process generating these specific/custom images without putting too much burden on the project. Having their definitions integrated into the NiFi code would make it easier to find them though, even if people had to build themselves.
Now, if the project thinks it's a good idea and is keen to integrate those images, I could change the approach and extend the maven builds with openshift profiles that would source their files from the already existing docker/dockerhub folders, as well as from the openshift one, but that would then only contain its specific files. The thing I'm struggling with though is that the way the different images are built (nifi, registry, toolkit) is not really consistent (for registry for example it does not seem to be a maven build at all, whereas in nifi you find two different things, dockerhub and dockermaven...), so I'm not sure how I should do this...
Another less intrusive option would be to simply add instructions/script in the parent folder and build directly from there. Like this, the build context would authorize to pick files from diffent subfolders and minimize duplication. This would also allow to provide only "recipes" for OpenShift, but not integrate the full build and have to determine where to put and how to name those images.
I'm open to any suggestion, whatever is the simple for the project and the users!

**Comment by kevdoran** [comment]
Hi @guimou @dmnewber,

This is a nice contribution and good discussion. I agree with you @guimou that there are real benefits to making NiFi images easier to deploy directly on OpenShift. I also agree with @dmnewber that it would be best not to duplicate the entire directory of files. While it may leave things more decoupled, it does increase the maintenance burden, as any any improvements to the default NiFi image would not also have to be done in more places (as we would probably want those same enhancements everywhere). We already have this problem today across docker hub and dockermaven variants, and plan to try to combine those so that they only differ one way, which is how the assembly artifacts are provided.

I've started collecting ideas for how to improve our project structure for Docker images produced from the NiFi repository here, and I've added some of the goals of this PR to that list: https://cwiki.apache.org/confluence/display/NIFI/NiFi+Docker+Container+Improvements

If this is not urgent, I would suggest we tackle some of those goals, such as image build consolidation and then figure out how to incorporate base image variants as part of that, to take another look at this. I haven't looked at the PR closely yet, but aside from basing on ubi8/OpenJDK11, are there any other significant changes from the current NiFi Dockerfile?

**Comment by guimou** [comment]
Hey @kevdoran! This is definitely not an urgent thing. And if it's the occasion to rethink other aspects of the container images building process that may be a good occasion and I'd be happy to help. Meanwhile I can keep a separate project (in my repo or another Red Hat repo) only for the OpenShift building recipes, not keeping the whole NiFi tree.
Apart from the source image and the installed packages that differ a little bit because it's not the same base distro, the changes are all related to the way the NiFi files are installed: ownership (for OpenShift GID has to be set to zero to support arbitrary UID) and access writes. In fact this could even be applied to the way the images are currently built without that many consequences (although it would require some more reflexion).

**Comment by kevdoran** [comment]
Thanks @guimou. Given that we want to offer a variety of base images and platform architectures, I think the first step would be restructuring things such that NiFi image is built in two steps: first to prep the base image, and then the nifi-specific bits. Ideally the nifi-bits would be the same regardless of the base image.

Once we get a better handle on that, it should be easier to tackle some of the other wish-list items for docker improvements as there will be fewer files to edit and maintain.

Keep an eye out on the wiki/dev mailing list/jira/PRs for opportunities to get involved if you are interested and available!

File: nifi-docker/openshift/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+EXPOSE 8080 8443 10000 8000
+
+WORKDIR ${NIFI_HOME}
+
+#OpenShift specific fixes
+RUN ${NIFI_BASE_DIR}/scripts/fix-permissions.sh ${NIFI_BASE_DIR}
+
+ENV HOME=${NIFI_HOME}
+
+USER 1001
```
Comment (cemremengu) [inline nifi-docker/openshift/Dockerfile:98]: This looks duplicates from above?

Comment (guimou) [reply nifi-docker/openshift/Dockerfile:98]: Good catch, this is already under USER 1001. I will remove it.

**Pull request closed without merging**